### DISCUSSION TITLE
[Feat]: watchFocus option

### DISCUSSION THREAD

**Discussion by S-Shingler** [davidjerleke/embla-carousel#936] [Ideas] [answered]
### Which variants of Embla Carousel are relevant to this feature request?

- [X] embla-carousel **(Core)**
- [ ] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [ ] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-fade
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

### Feature description

I would like to override or disable the default slide focus behaviour.

### Additional Context

- PR: #937.

I came across this bug when implementing accessibility for a belt. Essentially if the remaining slides was was less than the slides to scroll when I tab to the next slide it would jump back. This is happening because of this line in the slideFocus.ts file.

https://github.com/davidjerleke/embla-carousel/blob/387a3faeb9e8ccff915d6b53f30b9999d180388e/packages/embla-carousel/src/components/SlideFocus.ts#L46

User scrolls
<img width="863" alt="image" src="https://github.com/davidjerleke/embla-carousel/assets/48463809/e76b960d-60e5-46be-92c8-901de361b119">

In final scroll snap segment
<img width="874" alt="image" src="https://github.com/davidjerleke/embla-carousel/assets/48463809/c3302aa9-a54f-4bf9-b2d3-68e1484d8f4d">

Press tab and `scrollTo` method takes us back to previous scroll snap segment
<img width="815" alt="image" src="https://github.com/davidjerleke/embla-carousel/assets/48463809/95305829-7e31-494b-ba02-fa7151c7a683">

### Before submitting

- [X] I've made research efforts and searched the documentation
- [X] I've searched for existing issues
- [X] I agree to follow this project's Contributing Guidelines for feature requests

**Comment by davidjerleke** [comment]
Thanks @S-Shingler,

Would you mind creating a CodeSandbox and sharing steps to reproduce the problem you're describing?

**Reply by S-Shingler** [reply]
Hi @davidjerleke,

Apologies for the late response. Just had a chance to put something together now.

https://codesandbox.io/p/sandbox/embla-carousel-focus-bug-demo-72zqny

### Steps to reproduce

1. Scroll to the end of belt
2. Use `tab` or `shift + tab` to focus the eighth or ninth tile
3. Observe the belt jump back to the previous scroll snap group

**Reply by davidjerleke** [reply]
@S-Shingler thanks!

**Reply by davidjerleke** [reply]
Hi @S-Shingler,

I've tested your sandbox now and this is not a bug but rather a faulty setup. Check this [CodeSandbox](https://ck56pt.csb.app/) out and note the following:

- You can navigate through both **Embla** and the browser native **CSS scroll snap** with <kbd>tab</kbd> or <kbd>Shift</kbd> + <kbd>tab</kbd>.
- Both behave in a very similar way.

In your sandbox, you're changing tab indexes while slides enter and leave view and that's causing the weird behavior. This is because the focus event will trigger **before** the `slidesInView` event, so the behavior you're trying to opt out of has already happened.

### Summary
- The bug is caused by your code in your sandbox.
- Regardless, a way to opt out is a good idea so `watchFocus` will be added to the library.

Best,
David

**Reply by S-Shingler** [reply]
Hi @davidjerleke,

Thank you for getting back to me and taking the time to review my sandbox. 

Whilst your example is similar in behaviour it wouldn't solve my problem because I don't want the user to have to tab through every item before they can continue tabbing down the page. For example, what if I had 30 slides in my carousel? And then imagine I have a page full of carousels that's a lot of extra tabbing to progress through the page.

In terms of the `slidesInView` event I have to respectfully disagree. You can scroll to the end of the belt and observe the dom in the inspector, all the tab indexes are correct. It's the `scrollTo.index` method fired on slide focus that causes the visible slides to shift and trigger the `slidesInView` event again and update the tab indexes. Also, I know it’s not my code because the same behaviour (that I don’t want) happens in your example when you `shift` + `tab` with slide 10 focused. To be clear on my requirements, I don’t want the visible scroll snap group to change unless the user explicitly triggers the scroll methods (previous and next buttons) or drags/swipes the carousel.

Cheers,
Sam

**Reply by davidjerleke** [reply]
@S-Shingler thanks for your response.

> Whilst your example is similar in behaviour it wouldn't solve my problem because I don't want the user to have to tab through every item before they can continue tabbing down the page. For example, what if I had 30 slides in my carousel? And then imagine I have a page full of carousels that's a lot of extra tabbing to progress through the page.

@S-Shingler yes, I understand that what you want to achieve isn't possible without opting out of the Embla default behaviour. So as I said before, it's a good idea to add a `watchFocus` option to the library. But, the behaviour that you think is undesired is also how the browser native CSS scroll snaps work. By default, they force you to tab through every item.

> In terms of the slidesInView event I have to respectfully disagree. You can scroll to the end of the belt and observe the dom in the inspector, all the tab indexes are correct.

Well yes, the tab indexes may be correct but it's a race condition and the timing is wrong on your side. This is obvious because without your code it's working as expected.

> Also, I know it’s not my code because the same behaviour (that I don’t want) happens in your example when you shift + tab with slide 10 focused. 

Here's how Embla and the native browser is behaving when tabbing with <kbd>Tab</kbd> (tabbing to next):

https://github.com/user-attachments/assets/1eaffc5f-017a-4c63-bd6c-defe5bd64e06

Here's how Embla and the native browser is behaving when tabbing with <kbd>Shift</kbd> + <kbd>Tab</kbd> (tabbing to previous):

https://github.com/user-attachments/assets/c11bcf3a-74ef-41fd-bbf2-15e64a601d9c

Now **please note** that the only difference between the Embla and native browser behaviour is that the native behaviour seems to skip some slides. But it's not actually skipping them. They do receive focus too. It's just not changing the carousel position because it's considering some slides to be in view just because a single pixel of them are in view - hence it skips the scrolling.

All I'm saying that this isn't a bug but it's the Embla behaviour by design. So it's expected. Again, I understand that what you want to achieve isn't possible without opting out of the Embla default behaviour.

**Reply by S-Shingler** [reply]
Hi @davidjerleke,

Thank you for your response and taking the time to explain from your perspective, I can see where you're coming from now. I also appreciate that you acknowledge that the behaviour I want isn't possible without opting out of the default behaviour.

I'm really glad that watchFocus will be added to the library!

Cheers,
Sam

**Reply by davidjerleke** [reply, marked as answer]
This feature is available from [v8.2.0](https://github.com/davidjerleke/embla-carousel/releases/tag/v8.2.0) and up.
